ext/strings.go

```diff
@@ -485,6 +548,468 @@ func join(strs []string) (string, error) {
+		}
+		argFloatVal := arg.ConvertToType(types.DoubleType)
+		argFloat, ok := argFloatVal.Value().(float64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to float64", argFloatVal.Value())
+		}
+		var fmtStr string
+		if useUpper {
+			fmtStr = fmt.Sprintf("%%%dE", *precision)
+		} else {
+			fmtStr = fmt.Sprintf("%%%de", *precision)
+		}
+		matchedLocale, err := matchLanguage(locale)
+		if err != nil {
+			return "", fmt.Errorf("error matching locale: %w", err)
+		}
+		return message.NewPrinter(matchedLocale).Sprintf(fmtStr, argFloat), nil
+	}
+}
+
+func formatBinary(arg ref.Val, locale string) (string, error) {
+	argIntVal := arg.ConvertToType(types.IntType)
+	argInt, ok := argIntVal.Value().(int64)
+	if !ok {
+		return "", fmt.Errorf("could not convert \"%s\" to int64", argIntVal.Value())
+	}
+	// locale is intentionally unused as integers formatted as binary
+	// strings are locale-independent
+	return fmt.Sprintf("%b", argInt), nil
+}
+
+func formatHex(useUpper bool) clauseImpl {
+	return func(arg ref.Val, locale string) (string, error) {
+		fmtStr := "%x"
+		if useUpper {
+			fmtStr = "%X"
+		}
+		switch arg.Type() {
+		case types.StringType, types.BytesType:
+			if arg.Type() == types.BytesType {
+				return fmt.Sprintf(fmtStr, arg.Value().([]byte)), nil
+			} else {
+				return fmt.Sprintf(fmtStr, arg.Value().(string)), nil
+			}
+		case types.IntType:
+			argInt, ok := arg.Value().(int64)
+			if !ok {
+				return "", fmt.Errorf("could not convert \"%s\" to int64", arg.Value())
+			}
+			return fmt.Sprintf(fmtStr, argInt), nil
+		default:
+			return "", fmt.Errorf("only integers, byte buffers, and strings can be formatted as hex, was given %s", arg.Type().TypeName())
+		}
+	}
+}
+
+func formatOctal(arg ref.Val, locale string) (string, error) {
+	if arg.Type() != types.IntType {
+		return "", fmt.Errorf("octal clause can only be used on integers, was given %s", arg.Type().TypeName())
+	}
+	argIntVal := arg.ConvertToType(types.IntType)
+	argInt, ok := argIntVal.Value().(int64)
+	if !ok {
+		return "", fmt.Errorf("could not convert \"%s\" to int64", argIntVal.Value())
+	}
+	return fmt.Sprintf("%o", argInt), nil
+}
+
+func formatBool(arg ref.Val, locale string) (string, error) {
+	if arg.Type() != types.BoolType {
+		return "", fmt.Errorf("bool clause can only be used on booleans, was given %s", arg.Type().TypeName())
+	}
+	argBool := arg.Value().(bool)
+	if argBool {
+		return "true", nil
+	} else {
+		return "false", nil
+	}
+}
+
+func parseFormattingClause(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	i, precision, err := parsePrecision(lastStrIndex+1, formatStr)
+	if err != nil {
+		return -1, nil, fmt.Errorf("error while parsing precision: %w", err)
+	}
+	r := rune(formatStr[i])
+	i++
+	switch r {
+	case 's':
+		return i, formatString, nil
+	case 'd':
+		return i, formatDecimal, nil
+	case 'f':
+		return i, formatFixed(precision), nil
+	case 'e', 'E':
```

Comment:
I think it's worth pointing out that `%E` no-ops at present since `text/message`'s underlying `%E` implementation never uses the actual E character when doing scientific notation. We could remove `%E` since it seems like this is intended behavior in upstream.

(Hex strings are not passed through the i18n printer, so this isn't a concern for `%h`/`%H`.)

Reply:
If `%E` is non-functional, then it's better to remove it.

Reply:
Makes sense - I've removed %E.